### ISSUE TITLE
Spec caching issue

### ISSUE THREAD

**Issue by andyjagoe** [faros-ai/airbyte-connectors#436]
I’m creating a new connector and having what appear to be caching issues of my spec.json when I run source acceptance tests: `./scripts/source-acceptance-test.sh myproject-source`. It seems like maybe docker is caching my spec and not updating it when I change it? Does anyone know how to force docker to not do any caching when I build my connector?

I noticed this from test_core which appears to suggest spec is being cached:

https://github.com/airbytehq/airbyte/blob/f32c5fa8ca4e98dfbef549f1160f939dd90cb4ad/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py#L33

```
@pytest.fixture(name="actual_connector_spec")
def actual_connector_spec_fixture(request: BaseTest, docker_runner):
    if not request.instance.spec_cache:
        output = docker_runner.call_spec()
        spec_messages = filter_output(output, Type.SPEC)
        assert len(spec_messages) == 1, "Spec message should be emitted exactly once"
        spec = spec_messages[0].spec
        request.spec_cache = spec
    return request.spec_cache

```

Here are the errors I’m getting:

```
self = <source_acceptance_test.tests.test_core.TestSpec object at 0x4008847490>
actual_connector_spec = ConnectorSpecification(documentationUrl=AnyUrl('https://docs.faros.ai', scheme='https', host='docs.faros.ai', tld='ai'...malization=False, supportsDBT=False, supported_destination_sync_modes=None, authSpecification=None, advanced_auth=None)
connector_config = SecretDict(******)

    def test_config_match_spec(self, actual_connector_spec: ConnectorSpecification, connector_config: SecretDict):
        """Check that config matches the actual schema from the spec call"""
        # Getting rid of technical variables that start with an underscore
        config = {key: value for key, value in connector_config.data.items() if not key.startswith("_")}
    
        try:
            jsonschema.validate(instance=config, schema=actual_connector_spec.connectionSpecification)
        except jsonschema.exceptions.ValidationError as err:
>           pytest.fail(f"Config invalid: {err}")
E           Failed: Config invalid: 'test_url' is a required property
E           
E           Failed validating 'required' in schema:
E               {'$schema': 'http://json-schema.org/draft-07/schema#',
E                'additionalProperties': False,
E                'properties': {'test_url': {'examples': ['https://my-jenkins-server.example.com'],
E                                            'title': 'Jenkins Server URL Test',
E                                            'type': 'string'},
E                               'token': {'airbyte_secret': True,
E                                         'title': 'Jenkins Token',
E                                         'type': 'string'},
E                               'user': {'title': 'Jenkins User', 'type': 'string'}},
E                'required': ['test_url', 'user', 'token'],
E                'title': 'Jenkins Spec',
E                'type': 'object'}
E           
E           On instance:
E               {'server_url': 'https://eth-rinkeby.alchemyapi.io/v2/',
E                'token': '',
E                'user': 'chris'}

source_acceptance_test/tests/test_core.py:55: Failed

 test_core.py ⨯                                                    5% ▌         

―――――――――――――――――――― TestSpec.test_match_expected[inputs0] ―――――――――――――――――――――

self = <source_acceptance_test.tests.test_core.TestSpec object at 0x400a2dedc0>
connector_spec = ConnectorSpecification(documentationUrl=AnyUrl('https://docs.faros.ai', scheme='https', host='docs.faros.ai', tld='ai'...malization=False, supportsDBT=False, supported_destination_sync_modes=None, authSpecification=None, advanced_auth=None)
actual_connector_spec = ConnectorSpecification(documentationUrl=AnyUrl('https://docs.faros.ai', scheme='https', host='docs.faros.ai', tld='ai'...malization=False, supportsDBT=False, supported_destination_sync_modes=None, authSpecification=None, advanced_auth=None)

    def test_match_expected(self, connector_spec: ConnectorSpecification, actual_connector_spec: ConnectorSpecification):
        """Check that spec call returns a spec equals to expected one"""
        if connector_spec:
>           assert actual_connector_spec == connector_spec, "Spec should be equal to the one in spec.yaml or spec.json file"
E           AssertionError: Spec should be equal to the one in spec.yaml or spec.json file
E           assert equals failed
E             ConnectorSpecification(doc  ConnectorSpecification(doc 
E             umentationUrl=AnyUrl('http  umentationUrl=AnyUrl('http 
E             s://docs.faros.ai', scheme  s://docs.faros.ai', scheme 
E             ='https', host='docs.faros  ='https', host='docs.faros 
E             .ai', tld='ai', host_type=  .ai', tld='ai', host_type= 
E             'domain'), changelogUrl=No  'domain'), changelogUrl=No 
E             ne, connectionSpecificatio  ne, connectionSpecificatio ...
E             
E             ...Full output truncated (27 lines hidden), use '-vv' to show

source_acceptance_test/tests/test_core.py:62: AssertionError

```
In the example above, my actual spec.json uses the key “server_url” and not “test_url”. “test_url” was something I had tried previously, but now appears cached and re-building my connector does not update it.

**Comment by andyjagoe** [comment]
If anyone else has this problem, the solution is to clean the docker build cache: 
```
docker builder prune --all
```
[https://stackoverflow.com/questions/65405562/is-there-a-way-to-clean-docker-build-cache](https://stackoverflow.com/questions/65405562/is-there-a-way-to-clean-docker-build-cache)

**Comment by tovbinm** [comment]
Thanks for bringing this problem up and documenting the solution! @andyjagoe

**Comment by andyjagoe** [comment]
My pleasure @tovbinm Do you have any guidance on creating a new source with the faros-airbyte-cdk if my source needs to import ESM modules? My connecter is working fine when tested locally, but configuring it to work when run with `./scripts/source-acceptance-test.sh` is providing to be a huge headache!

**Comment by cjwooo** [comment]
IIUC, I believe the esModuleInterop flag allows importing of ESM modules. Are you experiencing a compilation error when pulling in a dependency?

**Comment by andyjagoe** [comment]
Thanks @cjwooo. The project compiles fine with `npm run build` and all my tests pass (though I had to do various transformations and use babel-jest) to get jest to load my ESM packages. The problem only appears when I try to run the connector with the source acceptance tests. I actually am using esModuleInterop and still getting the below errors.

There appear to be many posts about the challenge of importing an ESM package to typescript if you're going to be running it on a target that requires CommonJS. There's a new module type called node12 that might address this, but requires a dev build of typescript and may not be stable. Another potential solution is using ts-node. I'm still looking for something that a) works, and b) is reliable.

```
Found source web3-source
WEB3_SOURCE_TEST_CONFIG env var is not set
Building source image web3-source
[+] Building 1.9s (16/16) FINISHED                                                                                                                                     
 => [internal] load build definition from Dockerfile                                                                                                              0.0s
 => => transferring dockerfile: 37B                                                                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                 0.0s
 => => transferring context: 35B                                                                                                                                  0.0s
 => [internal] load metadata for docker.io/library/node:14-alpine                                                                                                 1.4s
 => [ 1/11] FROM docker.io/library/node:14-alpine@sha256:8845b4f88f64f8c56a39236648ba22946e806a6153c10911f77b70e5a2edb4ca                                         0.0s
 => [internal] load build context                                                                                                                                 0.2s
 => => transferring context: 27.11kB                                                                                                                              0.2s
 => CACHED [ 2/11] WORKDIR /home/node/airbyte                                                                                                                     0.0s
 => CACHED [ 3/11] RUN npm install -g npm@7 lerna tsc                                                                                                             0.0s
 => CACHED [ 4/11] COPY lerna.json .tsconfig.json package.json package-lock.json ./                                                                               0.0s
 => CACHED [ 5/11] RUN sed -i "/eslint\|husky\|jest\|lint-staged\|mockttp\|prettier/d" package.json                                                               0.0s
 => CACHED [ 6/11] COPY ./faros-airbyte-cdk ./faros-airbyte-cdk                                                                                                   0.0s
 => CACHED [ 7/11] COPY ./sources ./sources                                                                                                                       0.0s
 => CACHED [ 8/11] COPY ./destinations ./destinations                                                                                                             0.0s
 => CACHED [ 9/11] RUN lerna bootstrap --hoist                                                                                                                    0.0s
 => CACHED [10/11] RUN test -n "sources/web3-source" || (echo "'path' argument is not set, e.g --build-arg path=destinations/airbyte-faros-destination" && false  0.0s
 => CACHED [11/11] RUN ln -s "/home/node/airbyte/sources/web3-source/bin/main" "/home/node/airbyte/main"                                                          0.0s
 => exporting to image                                                                                                                                            0.0s
 => => exporting layers                                                                                                                                           0.0s
 => => writing image sha256:7e7ed0dade141a1d9de62b7f44e878375c8fb9dc99653971415f8a0ed84b60e5                                                                      0.0s
 => => naming to docker.io/library/web3-source                                                                                                                    0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
Running source acceptance tests against web3-source
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Test session starts (platform: linux, Python 3.9.11, pytest 6.2.5, pytest-sugar 0.9.4)
rootdir: /test_input
plugins: sugar-0.9.4, timeout-1.4.2
collecting ... 

―――――――――― ERROR at setup of TestSpec.test_config_match_spec[inputs0] ――――――――――

request = <SubRequest 'actual_connector_spec' for <Function test_config_match_spec[inputs0]>>
docker_runner = <source_acceptance_test.utils.connector_runner.ConnectorRunner object at 0x4007c0e3d0>

    @pytest.fixture(name="actual_connector_spec")
    def actual_connector_spec_fixture(request: BaseTest, docker_runner):
        if not request.instance.spec_cache:
>           output = docker_runner.call_spec()

source_acceptance_test/tests/test_core.py:34: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
source_acceptance_test/utils/connector_runner.py:68: in call_spec
    output = list(self.run(cmd=cmd, **kwargs))
source_acceptance_test/utils/connector_runner.py:106: in run
    for line in self.read(container, command=cmd):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'source_acceptance_test.utils.connector_runner.ConnectorRunner'>
container = <Container: e1efa6ab7f>, command = 'spec', with_ext = True

    @classmethod
    def read(cls, container: Container, command: str = None, with_ext: bool = True) -> Iterable[str]:
        """Reads connector's logs per line"""
        buffer = b""
        exception = ""
        line = ""
        for chunk in container.logs(stdout=True, stderr=True, stream=True, follow=True):
    
            buffer += chunk
            while True:
                # every chunk can include several lines
                found = buffer.find(b"\n")
                if found <= -1:
                    break
    
                line = buffer[: found + 1].decode("utf-8")
                if len(exception) > 0 or line.startswith("Traceback (most recent call last)"):
                    exception += line
                else:
                    yield line
                buffer = buffer[found + 1 :]
    
        if buffer:
            # send the latest chunk if exists
            line = buffer.decode("utf-8")
            if exception:
                exception += line
            else:
                yield line
        try:
            exit_status = container.wait()
            container.remove()
        except NotFound as err:
            logging.error(f"Waiting error: {err}, logs: {exception or line}")
            raise
        if exit_status["StatusCode"]:
            error = exit_status["Error"] or exception or line
            logging.error(f"Docker container was failed, " f'code {exit_status["StatusCode"]}, error:\n{error}')
            if with_ext:
>               raise ContainerError(
                    container=container,
                    exit_status=exit_status["StatusCode"],
                    command=command,
                    image=container.image,
                    stderr=error,
                )
E               docker.errors.ContainerError: Command 'spec' in image '<Image: 'web3-source:latest'>' returned non-zero exit status 1: }

source_acceptance_test/utils/connector_runner.py:152: ContainerError
---------------------------- Captured stdout setup -----------------------------
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output internal/modules/cjs/loader.js:1102\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=internal/modules/cjs/loader.js:1102\n; pos=0; lineno=1; colno=1)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output       throw new ERR_REQUIRE_ESM(filename, parentPath, packageJsonPath);\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 7 (char 6) (type=value_error.jsondecode; msg=Expecting value; doc=      throw new ERR_REQUIRE_ESM(filename, parentPath, packageJsonPath);\n; pos=6; lineno=1; colno=7)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output       ^\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 7 (char 6) (type=value_error.jsondecode; msg=Expecting value; doc=      ^\n; pos=6; lineno=1; colno=7)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output \n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 2 column 1 (char 1) (type=value_error.jsondecode; msg=Expecting value; doc=\n; pos=1; lineno=2; colno=1)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /home/node/airbyte/node_modules/@ceramicnetwork/http-client/lib/ceramic-http-client.js\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /home/node/airbyte/node_modules/@ceramicnetwork/http-client/lib/ceramic-http-client.js\n; pos=0; lineno=1; colno=1)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output require() of ES modules is not supported.\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=require() of ES modules is not supported.\n; pos=0; lineno=1; colno=1)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output require() of /home/node/airbyte/node_modules/@ceramicnetwork/http-client/lib/ceramic-http-client.js from /home/node/airbyte/sources/web3-source/lib/web3analytics.js is an ES module file as it is a .js file whose nearest parent package.json contains \"type\": \"module\" which defines all .js files in that package scope as ES modules.\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=require() of /home/node/airbyte/node_modules/@ceramicnetwork/http-client/lib/ceramic-http-client.js from /home/node/airbyte/sources/web3-source/lib/web3analytics.js is an ES module file as it is a .js file whose nearest parent package.json contains \"type\": \"module\" which defines all .js files in that package scope as ES modules.\n; pos=0; lineno=1; colno=1)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output Instead rename ceramic-http-client.js to end in .cjs, change the requiring code to use import(), or remove \"type\": \"module\" from /home/node/airbyte/node_modules/@ceramicnetwork/http-client/package.json.\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=Instead rename ceramic-http-client.js to end in .cjs, change the requiring code to use import(), or remove \"type\": \"module\" from /home/node/airbyte/node_modules/@ceramicnetwork/http-client/package.json.\n; pos=0; lineno=1; colno=1)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output \n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 2 column 1 (char 1) (type=value_error.jsondecode; msg=Expecting value; doc=\n; pos=1; lineno=2; colno=1)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at new NodeError (internal/errors.js:322:7)\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at new NodeError (internal/errors.js:322:7)\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at Module.load (internal/modules/cjs/loader.js:950:32)\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Module.load (internal/modules/cjs/loader.js:950:32)\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Function.Module._load (internal/modules/cjs/loader.js:790:12)\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at Module.require (internal/modules/cjs/loader.js:974:19)\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Module.require (internal/modules/cjs/loader.js:974:19)\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at require (internal/modules/cjs/helpers.js:101:18)\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at require (internal/modules/cjs/helpers.js:101:18)\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at Object.<anonymous> (/home/node/airbyte/sources/web3-source/lib/web3analytics.js:42:21)\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Object.<anonymous> (/home/node/airbyte/sources/web3-source/lib/web3analytics.js:42:21)\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at Module._compile (internal/modules/cjs/loader.js:1085:14)\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Module._compile (internal/modules/cjs/loader.js:1085:14)\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output     at Module.load (internal/modules/cjs/loader.js:950:32) {\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Module.load (internal/modules/cjs/loader.js:950:32) {\n; pos=4; lineno=1; colno=5)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output   code: 'ERR_REQUIRE_ESM'\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 3 (char 2) (type=value_error.jsondecode; msg=Expecting value; doc=  code: 'ERR_REQUIRE_ESM'\n; pos=2; lineno=1; colno=3)"}}
{"type": "LOG", "log": {"level": "WARN", "message": "Unable to parse connector's output }\n, error: 1 validation error for AirbyteMessage\n__root__\n  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=}\n; pos=0; lineno=1; colno=1)"}}
{"type": "LOG", "log": {"level": "ERROR", "message": "Docker container was failed, code 1, error:\n}\n"}}
------------------------------ Captured log setup ------------------------------
WARNING  root:connector_runner.py:111 Unable to parse connector's output internal/modules/cjs/loader.js:1102
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=internal/modules/cjs/loader.js:1102
; pos=0; lineno=1; colno=1)
WARNING  root:connector_runner.py:111 Unable to parse connector's output       throw new ERR_REQUIRE_ESM(filename, parentPath, packageJsonPath);
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 7 (char 6) (type=value_error.jsondecode; msg=Expecting value; doc=      throw new ERR_REQUIRE_ESM(filename, parentPath, packageJsonPath);
; pos=6; lineno=1; colno=7)
WARNING  root:connector_runner.py:111 Unable to parse connector's output       ^
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 7 (char 6) (type=value_error.jsondecode; msg=Expecting value; doc=      ^
; pos=6; lineno=1; colno=7)
WARNING  root:connector_runner.py:111 Unable to parse connector's output 
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 2 column 1 (char 1) (type=value_error.jsondecode; msg=Expecting value; doc=
; pos=1; lineno=2; colno=1)
WARNING  root:connector_runner.py:111 Unable to parse connector's output Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /home/node/airbyte/node_modules/@ceramicnetwork/http-client/lib/ceramic-http-client.js
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /home/node/airbyte/node_modules/@ceramicnetwork/http-client/lib/ceramic-http-client.js
; pos=0; lineno=1; colno=1)
WARNING  root:connector_runner.py:111 Unable to parse connector's output require() of ES modules is not supported.
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=require() of ES modules is not supported.
; pos=0; lineno=1; colno=1)
WARNING  root:connector_runner.py:111 Unable to parse connector's output require() of /home/node/airbyte/node_modules/@ceramicnetwork/http-client/lib/ceramic-http-client.js from /home/node/airbyte/sources/web3-source/lib/web3analytics.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=require() of /home/node/airbyte/node_modules/@ceramicnetwork/http-client/lib/ceramic-http-client.js from /home/node/airbyte/sources/web3-source/lib/web3analytics.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
; pos=0; lineno=1; colno=1)
WARNING  root:connector_runner.py:111 Unable to parse connector's output Instead rename ceramic-http-client.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /home/node/airbyte/node_modules/@ceramicnetwork/http-client/package.json.
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=Instead rename ceramic-http-client.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /home/node/airbyte/node_modules/@ceramicnetwork/http-client/package.json.
; pos=0; lineno=1; colno=1)
WARNING  root:connector_runner.py:111 Unable to parse connector's output 
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 2 column 1 (char 1) (type=value_error.jsondecode; msg=Expecting value; doc=
; pos=1; lineno=2; colno=1)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at new NodeError (internal/errors.js:322:7)
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at new NodeError (internal/errors.js:322:7)
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at Module.load (internal/modules/cjs/loader.js:950:32)
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Module.load (internal/modules/cjs/loader.js:950:32)
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at Function.Module._load (internal/modules/cjs/loader.js:790:12)
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at Module.require (internal/modules/cjs/loader.js:974:19)
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Module.require (internal/modules/cjs/loader.js:974:19)
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at require (internal/modules/cjs/helpers.js:101:18)
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at require (internal/modules/cjs/helpers.js:101:18)
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at Object.<anonymous> (/home/node/airbyte/sources/web3-source/lib/web3analytics.js:42:21)
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Object.<anonymous> (/home/node/airbyte/sources/web3-source/lib/web3analytics.js:42:21)
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at Module._compile (internal/modules/cjs/loader.js:1085:14)
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Module._compile (internal/modules/cjs/loader.js:1085:14)
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output     at Module.load (internal/modules/cjs/loader.js:950:32) {
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 5 (char 4) (type=value_error.jsondecode; msg=Expecting value; doc=    at Module.load (internal/modules/cjs/loader.js:950:32) {
; pos=4; lineno=1; colno=5)
WARNING  root:connector_runner.py:111 Unable to parse connector's output   code: 'ERR_REQUIRE_ESM'
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 3 (char 2) (type=value_error.jsondecode; msg=Expecting value; doc=  code: 'ERR_REQUIRE_ESM'
; pos=2; lineno=1; colno=3)
WARNING  root:connector_runner.py:111 Unable to parse connector's output }
, error: 1 validation error for AirbyteMessage
__root__
  Expecting value: line 1 column 1 (char 0) (type=value_error.jsondecode; msg=Expecting value; doc=}
; pos=0; lineno=1; colno=1)
ERROR    root:connector_runner.py:150 Docker container was failed, code 1, error:
}
                                                                   5% ▌         

----

<15 more similar errors here>

----


=========================== short test summary info ============================
FAILED ../../test_input/test_core.py::TestSpec::test_config_match_spec[inputs0]
FAILED ../../test_input/test_core.py::TestSpec::test_match_expected[inputs0]
FAILED ../../test_input/test_core.py::TestSpec::test_docker_env[inputs0] - do...
FAILED ../../test_input/test_core.py::TestSpec::test_oneof_usage[inputs0] - d...
FAILED ../../test_input/test_core.py::TestSpec::test_defined_refs_exist_in_json_spec_file[inputs0]
FAILED ../../test_input/test_core.py::TestSpec::test_oauth_flow_parameters[inputs0]
FAILED ../../test_input/test_core.py::TestConnection::test_check[inputs0] - d...
FAILED ../../test_input/test_core.py::TestConnection::test_check[inputs1] - d...
FAILED ../../test_input/test_core.py::TestDiscovery::test_discover[inputs0]
FAILED ../../test_input/test_core.py::TestDiscovery::test_defined_cursors_exist_in_schema[inputs0]
FAILED ../../test_input/test_core.py::TestDiscovery::test_defined_refs_exist_in_schema[inputs0]
FAILED ../../test_input/test_core.py::TestDiscovery::test_defined_keyword_exist_in_schema[inputs0-allOf]
FAILED ../../test_input/test_core.py::TestDiscovery::test_defined_keyword_exist_in_schema[inputs0-not]
FAILED ../../test_input/test_core.py::TestDiscovery::test_primary_keys_exist_in_schema[inputs0]
FAILED ../../test_input/test_core.py::TestBasicRead::test_read[inputs0] - doc...
FAILED ../../test_input/test_full_refresh.py::TestFullRefresh::test_sequential_reads[inputs0]
FAILED ../../test_input/test_incremental.py::TestIncremental::test_two_sequential_reads[inputs0]
FAILED ../../test_input/test_incremental.py::TestIncremental::test_state_with_abnormally_large_values[inputs0]

Results (13.31s):
       4 passed
       3 failed
         - ../airbyte/source_acceptance_test/source_acceptance_test/tests/test_core.py:160 TestConnection.test_check[inputs0]
         - ../airbyte/source_acceptance_test/source_acceptance_test/tests/test_core.py:160 TestConnection.test_check[inputs1]
         - ../airbyte/source_acceptance_test/source_acceptance_test/tests/test_core.py:183 TestDiscovery.test_discover[inputs0]
      15 error
web3-source failed source acceptance tests

```

**Comment by andyjagoe** [comment]
@cjwooo, here is one approach to getting something like this to work: https://stackoverflow.com/a/65163089. Node has some experimental loader hooks that can be set with `node --experimental-specifier-resolution=node index.js` or `node --loader ts-node/esm ./my-script.ts`. More info here: https://github.com/TypeStrong/ts-node/issues/1007

Where is node invoked in the faros-airbyte-cdk when running a typescript connector in docker for either source_acceptance_tests or production?

**Comment by cjwooo** [comment]
Interesting, I'm not too familiar with all the intricacies of Typescript compilation with ESM and CommonJS and so forth.

Node is invoked by both the source acceptance tests and Airbyte instances here https://github.com/faros-ai/airbyte-connectors/blob/main/Dockerfile#L20, which calls a source-specific executable (e.g. https://github.com/faros-ai/airbyte-connectors/blob/main/sources/bitbucket-source/bin/main)

**Comment by andyjagoe** [comment]
@cjwooo @tovbinm So to enable loading a pure ESM module in my source connector I needed to re-write example-source so that it's an ESM module and no longer CommonJS. I had to make some other changes as well. I renamed main to main.mjs so it will run as an ESM module and added the apk for coreutils in the Dockerfile so I can run the following shebang in main.mjs `#!/usr/bin/env -S ts-node --esm`. Once I get things cleaned up I'm going to see if I can run this under node instead of ts-node and will see if I can point you guys to a clean fork and/or put this in a separate issue so others can find it easily.

Here are some good references I used: [Pure ESM packages](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) and [Node ESM troubleshooting](https://docs.joshuatz.com/cheatsheets/node-and-npm/node-esm/)

Have you guys considered switching your baseline connector from CommonJS to ESM? As I've learned the hard way, ESM modules can easily import other ESM and CommonJS modules, but CommonJS modules cannot import pure ESM modules.
